infra/build/functions/update_build_status.py

```diff
@@ -233,19 +243,58 @@ def upload_log(build_id):
 # pylint: disable=no-member
 def update_status(event, context):
   """Entry point for cloud function to update build statuses and badges."""
-  del event, context  #unused
+  del context
+
+  if 'data' in event:
+    status_type = base64.b64decode(event['data']).decode()
+  else:
+    raise RuntimeError('No data')
+
+  if status_type == 'badges':
+    update_badges()
+    return
+
+  if status_type == 'fuzzing':
+    tag = build_project.FUZZING_BUILD_TAG
+    status_filename = FUZZING_STATUS_FILENAME
+  elif status_type == 'coverage':
+    tag = build_and_run_coverage.COVERAGE_BUILD_TAG
+    status_filename = COVERAGE_STATUS_FILENAME
+  else:
+    raise RuntimeError('Invalid type ' + status_type)
 
   with ndb.Client().context():
-    project_build_statuses = update_build_status(
-        build_project.FUZZING_BUILD_TAG, status_filename='status.json')
-    coverage_build_statuses = update_build_status(
-        build_and_run_coverage.COVERAGE_BUILD_TAG,
-        status_filename='status-coverage.json')
-
-    for project in Project.query():
-      if (project.name not in project_build_statuses or
-          project.name not in coverage_build_statuses):
-        continue
-
-      update_build_badges(project.name, project_build_statuses[project.name],
-                          coverage_build_statuses[project.name])
+    update_build_status(tag, status_filename)
+
+
+def load_status_from_gcs(filename):
+  """Load statuses from bucket."""
+  status_bucket = get_storage_client().get_bucket(STATUS_BUCKET)
+  status = json.loads(status_bucket.blob(filename).download_as_string())
+  result = {}
+
+  for project in status['projects']:
+    if project['history']:
+      result[project['name']] = project['history'][0]['success']
+
+  return result
+
+
+def update_badges():
+  """Update badges."""
+  project_build_statuses = load_status_from_gcs(FUZZING_STATUS_FILENAME)
+  coverage_build_statuses = load_status_from_gcs(COVERAGE_STATUS_FILENAME)
+
+  with concurrent.futures.ThreadPoolExecutor(max_workers=32) as executor:
```

Comment:
Do we really need that high parallelism here - even a common global var of 8 or 10 is fine.

Reply:
It makes things much faster here. The reason for the lower limit for the other case is due to QPS restrictions.